modules/src/applications/transfer/relay/send_transfer.rs

```diff
@@ -0,0 +1,116 @@
+use crate::applications::transfer::msgs::transfer::MsgTransfer;
+use crate::applications::transfer::packet::PacketData;
+use crate::applications::transfer::{Coin, IbcCoin, Source, TracePrefix};
+use crate::core::ics04_channel::handler::send_packet::send_packet;
+use crate::core::ics04_channel::packet::Packet;
+use crate::events::ModuleEvent;
+use crate::handler::{HandlerOutput, HandlerOutputBuilder};
+use crate::prelude::*;
+
+pub fn send_transfer<Ctx>(
+    ctx: &mut Ctx,
+    output: &mut HandlerOutputBuilder<()>,
+    msg: MsgTransfer,
+) -> Result<(), Error>
+where
+    Ctx: Ics20Context,
+{
+    if !ctx.is_send_enabled() {
+        return Err(Error::send_disabled());
+    }
+
+    let source_channel_end = ctx
+        .channel_end(&(msg.source_port.clone(), msg.source_channel))
+        .map_err(Error::ics04_channel)?;
+
+    let destination_port = source_channel_end.counterparty().port_id().clone();
+    let destination_channel = *source_channel_end
+        .counterparty()
+        .channel_id()
+        .ok_or_else(|| {
+            Error::destination_channel_not_found(msg.source_port.clone(), msg.source_channel)
+        })?;
+
+    // get the next sequence
+    let sequence = ctx
+        .get_next_sequence_send(&(msg.source_port.clone(), msg.source_channel))
+        .map_err(Error::ics04_channel)?;
+
+    let denom = match msg.token.clone() {
+        IbcCoin::Hashed(coin) => ctx
+            .get_denom_trace(&coin.denom)
+            .ok_or_else(Error::trace_not_found)?,
+        IbcCoin::Base(coin) => coin.denom.into(),
+    };
+
+    let sender = msg
+        .sender
+        .clone()
+        .try_into()
+        .map_err(|_| Error::parse_account_failure())?;
+
+    let prefix = TracePrefix::new(msg.source_port.clone(), msg.source_channel);
+    match denom.source_chain(&prefix) {
+        Source::Sender => {
+            let escrow_address =
+                ctx.get_channel_escrow_address(&msg.source_port, msg.source_channel)?;
+            ctx.send_coins(&sender, &escrow_address, &msg.token)?;
+        }
+        Source::Receiver => ctx.burn_coins(&sender, &msg.token)?,
+    }
+
+    let data = {
+        let data = PacketData {
+            token: Coin {
+                denom,
+                amount: msg.token.amount(),
+            },
+            sender: msg.sender.clone(),
+            receiver: msg.receiver.clone(),
+        };
+        serde_json::to_vec(&data).expect("PacketData's infallible Serialize impl failed")
+    };
+
+    let packet = Packet {
+        sequence,
+        source_port: msg.source_port,
+        source_channel: msg.source_channel,
+        destination_port,
+        destination_channel,
+        data,
+        timeout_height: msg.timeout_height,
+        timeout_timestamp: msg.timeout_timestamp,
+    };
+
+    let HandlerOutput {
+        result,
+        log,
+        events,
+    } = send_packet(ctx, packet).map_err(Error::ics04_channel)?;
```

Comment:
The assumption is that the chain will revert any changes made on error, right? Otherwise, if `send_packet()` fails, then we will have escrowed/burned the tokens without actually sending them.

Reply:
Right, I documented this in a comment -> 27a5d74733da424bc470d75024afec54b0900529